File: src/data/courses/dsa/merge-sort/merge-sort-core/count-comparisons-merge-sort/code/cpp/CountComparisonsMergeSort.cpp
int merge(vector<int>& arr, int l, int m, int r) {
    vector<int> L(arr.begin() + l, arr.begin() + m + 1);
    vector<int> R(arr.begin() + m + 1, arr.begin() + r + 1);
    int i = 0, j = 0, k = l, count = 0;
    while (i < L.size() && j < R.size()) {
        count++;
        arr[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];
    }
    while (i < L.size()) arr[k++] = L[i++];
    while (j < R.size()) arr[k++] = R[j++];
    return count;
}

int mergeSort(vector<int>& arr, int l, int r) {
    if (l >= r) return 0;
    int m = l + (r - l) / 2;
    return mergeSort(arr, l, m) + mergeSort(arr, m + 1, r) + merge(arr, l, m, r);
}

File: src/data/courses/dsa/merge-sort/merge-sort-core/count-comparisons-merge-sort/code/python/count_comparisons_merge_sort.py
def merge_sort(arr, l, r):
    if l >= r:
        return 0
    m = l + (r - l) // 2
    count = merge_sort(arr, l, m) + merge_sort(arr, m + 1, r)
    return count + merge(arr, l, m, r)

def merge(arr, l, m, r):
    L = arr[l:m+1]
    R = arr[m+1:r+1]
    i = j = 0
    k = l
    count = 0
    while i < len(L) and j < len(R):
        count += 1
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        k += 1
    while i < len(L):
        arr[k] = L[i]
        i += 1
        k += 1
    while j < len(R):
        arr[k] = R[j]
        j += 1
        k += 1
    return count